### PR TITLE
Syndesis API discovery by 3scale fix

### PR DESCRIPTION
**Pull request by jsafarik**
//test: `@3scale`

### CODE REVIEW

**Comment by syndesisqe** [comment]
Following scenarios failed:
```
3scale integration | 3scale discovery annotations 
```

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Checkstyle verification failed with:
```
utilities/src/main/java/io/syndesis/qe/utils/
  HttpUtils.java:120:65: '?' is not preceded with whitespace. [WhitespaceAround]
```

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Following scenarios failed:
```
Connection - CRUD | Validate all credentials 
3scale integration | 3scale discovery annotations 
```

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

File: ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java [hunk 1]
```diff
@@ -932,9 +896,20 @@ private boolean isStringInUrl(String toFind, int timeoutSeconds) {
...
-    @When("^set 3scale discovery variable to \"([^\"]*)\"")
+    @When("^enable 3scale discovery with url \"([^\"]*)\"")
     public void set3scaleEnvVar(String value) {
-        OpenShiftUtils.updateEnvVarInDeploymentConfig("syndesis-server", "CONTROLLERS_EXPOSE_VIA3SCALE", value);
+        String feature = "\"managementUrlFor3scale\":\"" + value + "\"";
+        if ("".equals(value)) {
+            feature = "";
+        }
+        HTTPResponse rsp = OpenShiftUtils.invokeApi(
+            HttpUtils.Method.PATCH,
```
Comment (asmigala) [inline ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java]: You can now use methods in the `SyndesisTemplate` class to edit the CR:
https://github.com/syndesisio/syndesis-qe/blob/master/utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java#L90-L96

The `getDeployedCr` will return the CR as a hierarchy of Maps, you can modify that and pass it `editCr`. Working with the maps is not super convenient (lots of type casting and `computeIfAbsent`), but refactoring that into something more pleasant is on my radar.

Comment (jsafarik) [reply ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java]: ok thanks, will give it a look and redo it :)

**Comment by mcada** [comment]
@jsafarik what is the state of this PR?

**Comment by syndesisqe** [comment]
Checkstyle verification failed with:
```
ui-tests/src/test/java/io/syndesis/qe/steps/
  CommonSteps.java:910:24: Unnecessary parentheses around assignment right-hand side. [UnnecessaryParentheses]
```

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

File: utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java [hunk 2]
```diff
@@ -82,21 +82,21 @@ private static void deployUsingOperator() {
...
-        getSyndesisCrClient().delete(namespace, name);
+    public static void deleteCr() {
+        getSyndesisCrClient().delete(TestConfiguration.openShiftNamespace(), "app");
     }
 
-    public static Set<String> getCrNames(String namespace) {
+    public static Set<String> getCrNames() {
         final Set<String> names = new HashSet<>();
-        Map<String, Object> crs = getSyndesisCrClient().list(namespace);
+        Map<String, Object> crs = getSyndesisCrClient().list(TestConfiguration.openShiftNamespace());
```
Comment (asmigala) [inline utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java]: @avano we agreed with @jsafarik that hard-coding the "app" name and always getting the namespace from the test config is ok here, wdyt?

Comment (asmigala) [reply utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java]: And I obviously don't know how the new multiline comments work, this should have been for all the changes in this file

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

File: utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java (hunk 2 above)
Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java]: maybe I would extract it to variable or constant, but other than that I have no objections

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

File: utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java (hunk 2 above)
Comment (jsafarik) [reply utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java]: Changed it to use a constant

Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java]: I know, but `This pull request is still a work in progress` :dancer:

File: ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java [hunk 3]
```diff
@@ -943,9 +903,24 @@ private boolean isStringInUrl(String toFind, int timeoutSeconds) {
...
+
+        Map cr = SyndesisTemplate.getDeployedCr();
+        Map features = (Map<String, Object>) ((Map<String, Object>) ((Map<String, Object>) ((Map<String, Object>) cr.get("spec")).get("components"))
+            .get("server")).get("features");
+
+        if (!"".equals(value)) {
+            features.put("managementUrlFor3scale", value);
+        } else {
+            features.remove("managementUrlFor3scale");
+        }
```
Comment (asmigala) [inline ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java]: Can't say I'm super excited about this - the step is is called `enable 3scale discovery` and here we have a magic value that will disable it. Maybe add a separate step for disabling?

File: utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java (hunk 2 above)
Comment (jsafarik) [reply utilities/src/main/java/io/syndesis/qe/templates/SyndesisTemplate.java]: yup, didn't noticed at first, thanks :smile:

File: ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java (hunk 3 above)
Comment (jsafarik) [reply ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java]: Indeed, it is not the best solution, refactored it so there are two public methods for enabling and disabling

File: ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java [hunk 4]
```diff
@@ -1010,5 +998,43 @@ public void saveAfterTime() {
...
+
+    private static class Element {
+        public static final By LOGIN_BUTTON = By.className("btn");
+        public static final By NAVIGATION_PANEL = By.className("pf-c-nav");
+        public static final By NAVIGATION_USER_DROPDOWN = By.cssSelector("[data-testid=app-top-menu-user-dropdown]");
+
+        public static final By navigationLink(String title) {
+            return By.cssSelector(String.format("a[data-testid=\"ui-%s\"]", title.toLowerCase().replaceAll("[\\s_]", "-")));
+        }
+    }
```
Comment (asmigala) [inline ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java]: This might be a misconfiguration in your IDE, should not happen

Comment (jsafarik) [reply ui-tests/src/test/java/io/syndesis/qe/steps/CommonSteps.java]: Don't think I've enabled or recofigured something like that in my IDE, but it seemed strange to me as well. Have to go thru settings and see if I can switch this behavior off.

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Pull request merged**